coderd/workspaceagentsrpc_internal_test.go

```diff
@@ -0,0 +1,376 @@
+package coderd
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/coder/coder/v2/coderd/util/ptr"
+
+	"github.com/golang/mock/gomock"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/require"
+	"nhooyr.io/websocket"
+
+	"cdr.dev/slog"
+	"cdr.dev/slog/sloggers/slogtest"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbmock"
+	"github.com/coder/coder/v2/coderd/database/dbtime"
+	"github.com/coder/coder/v2/testutil"
+)
+
+func TestAgentWebsocketMinder_ContextCancel(t *testing.T) {
+	t.Parallel()
+	ctx := testutil.Context(t, testutil.WaitShort)
+	now := dbtime.Now()
+	fConn := &fakePingerCloser{}
+	ctrl := gomock.NewController(t)
+	mDB := dbmock.NewMockStore(ctrl)
+	fUpdater := &fakeUpdater{}
+	logger := slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+	agent := database.WorkspaceAgent{
+		ID: uuid.New(),
+		FirstConnectedAt: sql.NullTime{
+			Time:  now.Add(-time.Minute),
+			Valid: true,
+		},
+	}
+	build := database.WorkspaceBuild{
+		ID:          uuid.New(),
+		WorkspaceID: uuid.New(),
+	}
+	replicaID := uuid.New()
+
+	uut := &agentWebsocketMinder{
+		apiCtx:            ctx,
+		workspaceAgent:    agent,
+		workspaceBuild:    build,
+		conn:              fConn,
+		db:                mDB,
+		replicaID:         replicaID,
+		updater:           fUpdater,
+		logger:            logger,
+		pingPeriod:        testutil.IntervalFast,
+		disconnectTimeout: testutil.WaitShort,
+	}
+	uut.init()
+
+	connected := mDB.EXPECT().UpdateWorkspaceAgentConnectionByID(
+		gomock.Any(),
+		connectionUpdate(agent.ID, replicaID),
+	).
+		AnyTimes().
+		Return(nil)
+	mDB.EXPECT().UpdateWorkspaceAgentConnectionByID(
+		gomock.Any(),
+		connectionUpdate(agent.ID, replicaID, withDisconnected()),
+	).
+		After(connected).
+		Times(1).
+		Return(nil)
+	mDB.EXPECT().GetLatestWorkspaceBuildByWorkspaceID(gomock.Any(), build.WorkspaceID).
+		AnyTimes().
+		Return(database.WorkspaceBuild{ID: build.ID}, nil)
+
+	closeCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	done := make(chan struct{})
+	go func() {
+		uut.mind(closeCtx)
+		close(done)
+	}()
+	// wait a couple intervals, but not long enough for a disconnect
+	time.Sleep(3 * testutil.IntervalFast)
```

Comment:
Would it make sense to use a channel here for synchronization/triggering a few intervals immediately instead of guessing we've waited long enough?

Reply:
What I'm trying to test here is that the monitor runs normally without closing.  Since golang doesn't support monkey patching in a generic way, we'd have to insert a lot of plumbing into the struct, which just doesn't seem worthwhile for shaving a few milliseconds off this test.

Reply:
Alright, as long as it doesn't result in flakes I'm good with that reasoning 👍🏻